src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/toolkit/taglets/snippet/Parser.java

```diff
@@ -0,0 +1,496 @@
+                    var type = attributes.get("type", Attribute.Valued.class);
+
+                    String typeValue = type.isPresent() ? type.get().value() : "bold";
+
+                    AddStyle a = new AddStyle(new Style.Name(typeValue),
+                            createRegexPattern(substring, regex,
+                                    t.lineSourceOffset + t.markupLineOffset),
+                            text.subText(t.start(), t.end()));
+                    actions.add(a);
+                }
+                case "start" -> {
+                    var region = attributes.get("region", Attribute.Valued.class)
+                            .orElseThrow(() -> new ParseException("Unnamed start",
+                                    t.lineSourceOffset + t.markupLineOffset + t.nameLineOffset));
+                    String regionValue = region.value();
+                    if (regionValue.isBlank()) {
+                        throw new ParseException("Blank region name",
+                                t.lineSourceOffset + t.markupLineOffset + region.valueStartPosition());
+                    }
+                    if (t.attributes().size() != 1) {
+                        throw new ParseException("Unexpected attributes",
+                                t.lineSourceOffset + t.markupLineOffset + t.nameLineOffset);
+                    }
+                    actions.add(new Bookmark(region.value(), text.subText(t.start(), t.end() - 1)));
+                }
+            }
+        }
+
+        return new Result(text, actions);
+    }
+
+    private Pattern createRegexPattern(Optional<Attribute.Valued> substring,
+                                       Optional<Attribute.Valued> regex,
+                                       int offset) throws ParseException {
+        return createRegexPattern(substring, regex, "(?s).+", offset);
+    }
+
+    private Pattern createRegexPattern(Optional<Attribute.Valued> substring,
+                                       Optional<Attribute.Valued> regex,
+                                       String defaultRegex,
+                                       int offset) throws ParseException {
+        Pattern pattern;
+        if (substring.isPresent()) {
+            // this Pattern.compile *cannot* throw an exception
+            pattern = Pattern.compile(Pattern.quote(substring.get().value()));
+        } else if (regex.isEmpty()) {
+            // this Pattern.compile *should not* throw an exception
+            pattern = Pattern.compile(defaultRegex);
+        } else {
+            // Unlike string literals in Java source, attribute values in
+            // snippet markup do not use escapes. This is why indices of
+            // characters in the regex pattern directly map to their
+            // corresponding positions in snippet source.
+            final String value = regex.get().value();
+            assert value.equals(Pattern.compile(value).pattern()) : value;
```

Comment:
What is the purpose of this assertion, and when would it ever fail?

Reply:
I think that this assertion is misguided. The assertion I wanted to make should be a test, which I will provide shortly. Thanks for directing my attention to it.

The purpose of that assertion was to encode the following factual statement: the value of the regex attribute is WYSIWYG. That is, javadoc interprets that value as is: characters or sequences thereof are treated literally. Compare this with Java source or jshell, where escape sequences are recognized:

```
jshell> Pattern.compile("name\t=\t\"value\"")
$1 ==> name	=	"value"
```

Why is it important? Because of error messages. Knowing how a regex is represented in source allows javadoc to map the index passed in an instance of `java.util.regex.PatternSyntaxException` back to regex source so as to adjust the caret position in a diagnostic message should the regex be invalid.